packages/next/src/build/webpack/loaders/action-utils.ts

```diff
@@ -0,0 +1,70 @@
+const key = crypto.subtle.generateKey(
+  {
+    name: 'AES-GCM',
+    length: 256,
+  },
+  true,
+  ['encrypt', 'decrypt']
+)
+
+async function encrypt(salt: string, data: ArrayBuffer) {
+  const iv = new TextEncoder().encode(salt)
+  return crypto.subtle.encrypt(
+    {
+      name: 'AES-GCM',
+      iv,
+    },
+    await key,
+    data
+  )
+}
+
+async function decrypt(salt: string, data: ArrayBuffer) {
+  const iv = new TextEncoder().encode(salt)
+  return crypto.subtle.decrypt(
+    {
+      name: 'AES-GCM',
+      iv,
+    },
+    await key,
+    data
+  )
+}
+
+function arrayBufferToString(buffer: ArrayBuffer) {
+  let binary = ''
+  let bytes = new Uint8Array(buffer)
+  let len = bytes.byteLength
+  for (let i = 0; i < len; i++) {
+    binary += String.fromCharCode(bytes[i])
+  }
+  return binary
+}
+
+function stringToArrayBuffer(binary: string) {
+  let len = binary.length
+  let bytes = new Uint8Array(len)
+  for (let i = 0; i < len; i++) {
+    bytes[i] = binary.charCodeAt(i)
+  }
+  return bytes.buffer
+}
+
+export async function decodeActionBoundArg(
+  actionId: string,
+  arg: Promise<string>
+) {
+  const decoded = await decrypt(
+    '__next_action__' + actionId,
+    stringToArrayBuffer(atob(await arg))
```

Comment:
Atob is legacy. `Buffer.from(data, 'base64')` is recommended, assuming this code runs on Node.js

Reply:
It's unfortunate that this also runs on the Edge runtime (similar to browser/worker environments).